File: time-to-go-home.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# 
# 始業時刻を与えるか現在時刻から規定の終業時刻を表示するスクリプト

import argparse
import re
import sys
import json
from datetime import datetime, timedelta

work_time = 8.5

def main():
    parser = argparse.ArgumentParser(description="show time to go home")
    parser.add_argument("-s", "--starttime", type=str, help="start time")
    parser.add_argument("-r", "--referencetime", type=str, help="referencetime for overtime calculation (test purpose)")
    parser.add_argument("-o", "--overtime", action="store_true", help="show over time (use with starttime option)")
    parser.add_argument("-j", "--json", action="store_true", help="output json")
    parser.add_argument("-d", "--debug", action="store_true", help="use debug")
    args = parser.parse_args()

    today = datetime.today()
    today_str = today.strftime("%Y-%m-%d")

    if args.starttime : 
        pattern = r"^\d{1,2}:\d{1,2}$"

        if not re.match(pattern, args.starttime):
            sys.stderr.write(f"error : incorrect time format. only 'hh:mm' format is acceptable: {args.starttime}\n")
            return(-1)

        time_str = f"{today_str} {args.starttime}"
        start_time = datetime.strptime(time_str, "%Y-%m-%d %H:%M")
    else :
        start_time = datetime.now()

    end_time = start_time + timedelta(hours=work_time)

    over_time = "00:00"
    now_time = None
    if args.overtime :
        # overtime計算用の時刻を決定
        if args.referencetime:
            # referencetimeはhh:mm形式
            pattern = r"^\d{1,2}:\d{1,2}$"
            if not re.match(pattern, args.referencetime):
                sys.stderr.write(f"error : incorrect referencetime format. only 'hh:mm' format is acceptable: {args.referencetime}\n")
                return(-1)
            ref_time_str = f"{today_str} {args.referencetime}"
            now_time = datetime.strptime(ref_time_str, "%Y-%m-%d %H:%M")
        else:
            now_time = datetime.now()
        # Over time計算（マイナスも対応）
        time_diff = now_time - end_time
        total_minutes = int(time_diff.total_seconds() // 60)
        sign = "-" if total_minutes < 0 else ""
        total_minutes = abs(total_minutes)
        diff_hour = total_minutes // 60
        diff_min = total_minutes % 60
        over_time = f"{sign}{diff_hour:02}:{diff_min:02}"
    else:
        now_time = datetime.now()

    if args.debug :
        debug_msg = f"Start time = {start_time.strftime('%H:%M')}, End time = {end_time.strftime('%H:%M')}, Reference time = {now_time.strftime('%H:%M')}, Over time = {over_time}\n"
        sys.stderr.write(debug_msg)

    if args.json : 
        result = {
            "items": [
                {
                    "title": "time to go home",
                    "start-time": start_time.strftime("%H:%M"),
                    "end-time": end_time.strftime("%H:%M"),
                    "over-time": over_time if args.overtime else None
                }
            ]
        }
        print(json.dumps(result))        
    else :
        if args.overtime:
            print(f"EndTime={end_time.strftime('%H:%M')}/OverTime={over_time}")
        else:
            print(f"EndTime={end_time.strftime('%H:%M')}")

    return(0)

if __name__ == "__main__":
    main()

File: time-to-go-home.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

#
# Show time to go home based on start time
# Calculates end time and overtime based on an 8.5-hour work day
#
# Usage:
#   time-to-go-home.rb [options]
#
# Options:
#   -s, --starttime HH:MM       Start time (default: current time)
#   -r, --referencetime HH:MM   Reference time for overtime calculation (for testing)
#   -o, --overtime              Show overtime
#   -j, --json                  Output as JSON
#   -d, --debug                 Show debug info
#   -h, --help                  Show this help
#

require 'optparse'
require 'json'
require 'time'

WORK_HOURS = 8.5

class WorkTimeCalculator
  def initialize(start_time:, reference_time: nil, overtime: false, json_output: false, debug: false)
    @start_time = start_time
    @reference_time = reference_time
    @overtime = overtime
    @json_output = json_output
    @debug = debug
  end

  def end_time
    @end_time ||= @start_time + (WORK_HOURS * 3600)
  end

  def format_overtime
    now = @reference_time || Time.now
    total_minutes = ((now - end_time) / 60).floor
    sign = total_minutes < 0 ? '-' : ''
    total_minutes = total_minutes.abs
    format('%s%02d:%02d', sign, total_minutes / 60, total_minutes % 60)
  end

  def run
    ot = @overtime ? format_overtime : nil
    now = @reference_time || Time.now

    if @debug
      warn "Start time = #{@start_time.strftime('%H:%M')}, End time = #{end_time.strftime('%H:%M')}, " \
           "Reference time = #{now.strftime('%H:%M')}, Over time = #{ot || '(none)'}"
    end

    if @json_output
      result = {
        'items' => [
          {
            'title'      => 'time to go home',
            'start-time' => @start_time.strftime('%H:%M'),
            'end-time'   => end_time.strftime('%H:%M'),
            'over-time'  => ot
          }
        ]
      }
      puts result.to_json
    elsif @overtime
      puts "EndTime=#{end_time.strftime('%H:%M')}/OverTime=#{ot}"
    else
      puts "EndTime=#{end_time.strftime('%H:%M')}"
    end
  end
end

def parse_hhmm(str, label)
  unless str.match?(/^\d{1,2}:\d{2}$/)
    warn "error: incorrect #{label} format. only 'hh:mm' format is acceptable: #{str}"
    exit 1
  end

  today = Time.now.strftime('%Y-%m-%d')
  Time.strptime("#{today} #{str}", '%Y-%m-%d %H:%M')
end

def main
  options = { starttime: nil, referencetime: nil, overtime: false, json: false, debug: false }

  OptionParser.new do |opts|
    opts.banner = 'Usage: time-to-go-home.rb [options]'

    opts.on('-s', '--starttime HH:MM', 'Start time (default: current time)') do |v|
      options[:starttime] = v
    end

    opts.on('-r', '--referencetime HH:MM', 'Reference time for overtime calculation') do |v|
      options[:referencetime] = v
    end

    opts.on('-o', '--overtime', 'Show overtime') do
      options[:overtime] = true
    end

    opts.on('-j', '--json', 'Output as JSON') do
      options[:json] = true
    end

    opts.on('-d', '--debug', 'Show debug info') do
      options[:debug] = true
    end

    opts.on_tail('-h', '--help', 'Show this help') do
      puts opts
      exit 0
    end
  end.parse!

  start_time = options[:starttime] ? parse_hhmm(options[:starttime], 'starttime') : Time.now
  ref_time   = options[:referencetime] ? parse_hhmm(options[:referencetime], 'referencetime') : nil

  WorkTimeCalculator.new(
    start_time:    start_time,
    reference_time: ref_time,
    overtime:       options[:overtime],
    json_output:    options[:json],
    debug:          options[:debug]
  ).run
end

main if $PROGRAM_NAME == __FILE__
